mm2src/coins/z_coin/storage/walletdb/mod.rs

```diff
@@ -33,27 +46,115 @@ pub struct WalletDbShared {
     ticker: String,
 }
 
+/// Calculates the starting block height based on a given date and the current block height.
+///
+/// # Arguments
+/// * `date`: The date in seconds representing the desired starting date.
+/// * `current_block_height`: The current block height at the time of calculation.
+///
+#[cfg(not(target_arch = "wasm32"))]
+fn calculate_starting_height_from_date(date: u32, current_block_height: u64) -> Result<u32, String> {
+    let blocks_prod_per_day = 24 * 60;
+    let current_time_s = now_sec();
+
+    let date = date as u64;
+    if current_time_s < date {
+        return Err("sync_param_date must be earlier then current date".to_string());
+    };
+
+    let secs_since_date = current_time_s - date;
+    let days_since_date = (secs_since_date / 86400) - 1;
+    let blocks_to_sync = (days_since_date * blocks_prod_per_day) + blocks_prod_per_day;
+    let block_to_sync_from = current_block_height - blocks_to_sync;
+
+    Ok(block_to_sync_from as u32)
```

Comment:
We are recasting here as u32, I think we should avoid casting and recasting inside this function (make it return u64) and we should cast outside the function since we will make it a common function.

Reply:
done